controller/internal/enforcer/envoyproxy/authz/server.go

```diff
@@ -0,0 +1,501 @@
+	if err := s.server.Serve(lis); err != nil {
+		zap.L().Error("gRPC server for ext_authz failed", zap.String("puID", s.puID), zap.Error(err), zap.String("direction", s.direction.String()))
+	}
+	zap.L().Debug("stopped serving gRPC for ext_authz server", zap.String("puID", s.puID), zap.String("direction", s.direction.String()))
+}
+
+// Stop calls the function with the same name on the backing gRPC server
+func (s *Server) Stop() {
+	s.server.Stop()
+}
+
+// GracefulStop calls the function with the same name on the backing gRPC server
+func (s *Server) GracefulStop() {
+	s.server.GracefulStop()
+}
+
+// Check implements the AuthorizationServer interface
+func (s *Server) Check(ctx context.Context, checkRequest *ext_authz_v2.CheckRequest) (*ext_authz_v2.CheckResponse, error) {
+	switch s.direction {
+	case IngressDirection:
+		return s.ingressCheck(ctx, checkRequest)
+	case EgressDirection:
+		return s.egressCheck(ctx, checkRequest)
+	default:
+		return nil, fmt.Errorf("direction: %s", s.direction)
+	}
+}
+
+// ingressCheck implements the AuthorizationServer for ingress connections
+func (s *Server) ingressCheck(ctx context.Context, checkRequest *ext_authz_v2.CheckRequest) (*ext_authz_v2.CheckResponse, error) {
+	// TODO: needs to be removed before we merge: this exposes secret data, and must never be in real logs - even in the debug case
+	zap.L().Debug("ext_authz ingress: checkRequest", zap.String("puID", s.puID), zap.String("checkRequest", checkRequest.String()))
+
+	// get the PU context
+	pctxRaw, err := s.puContexts.Get(s.puID)
+	if err != nil {
+		zap.L().Error("ext_authz ingress: failed to get PU context", zap.String("puID", s.puID), zap.Error(err))
+		return createDeniedCheckResponse(rpc.INTERNAL, envoy_type.StatusCode_InternalServerError, "failed to get PU context"), nil
+	}
+	pctx, ok := pctxRaw.(*pucontext.PUContext)
+	if !ok {
+		zap.L().Error("ext_authz ingress: PU context has the wrong type", zap.String("puID", s.puID), zap.String("puContextType", fmt.Sprintf("%T", pctxRaw)))
+		return createDeniedCheckResponse(rpc.INTERNAL, envoy_type.StatusCode_InternalServerError, "PU context has the wrong type"), nil
+	}
+
+	// extract our headers
+	var aporetoKey, aporetoAuth, sourceAdress string
+	var source, dest *ext_authz_v2.AttributeContext_Peer
+	var httpReq *ext_authz_v2.AttributeContext_HttpRequest
+	attrs := checkRequest.GetAttributes()
+	if attrs != nil {
+		source = attrs.GetSource()
+		dest = attrs.GetDestination()
+
+		if source != nil {
+			if addr := source.GetAddress(); addr != nil {
+				if sockAddr := addr.GetSocketAddress(); sockAddr != nil {
+					sourceAdress = sockAddr.GetAddress()
+				}
+			}
+		}
+
+		if request := attrs.GetRequest(); request != nil {
+			httpReq = request.GetHttp()
+			if httpReq != nil {
+				httpReqHeaders := httpReq.GetHeaders()
+				if httpReqHeaders != nil {
+					aporetoKey, _ = httpReqHeaders[aporetoKeyHeader]
+					aporetoAuth, _ = httpReqHeaders[aporetoAuthHeader]
+				}
+			}
+		}
+	}
+
```

Comment:
if attrs is nil it flows here. Is that what is needed. Shouldn't we just return for nil attributes?

Reply:
if attrs in nil, it will not have extracted the headers and we will fail a couple of lines below